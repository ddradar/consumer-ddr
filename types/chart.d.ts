export type PlayStyle =
  | 'SINGLE'
  | 'DOUBLE'
  | 'COUPLE'
  | '6-PANELS'
  | '3-PANELS'
  | 'STEP BATTLE'

export type Chart = {
  songId: string
  playStyle: PlayStyle
  difficulty: number
  level: number | '10+'
}
