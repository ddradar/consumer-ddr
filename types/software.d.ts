type Region = 'JP' | 'US' | 'EU' | 'None'

type Platform =
  | 'Play Station'
  | 'Play Station 2'
  | 'Dreamcast'
  | 'GAME BOY COLOR'
  | 'Wii'

export type Software = {
  id: string
  name: string
  region: Region
  platform: Platform
  launched: Date | null
}
