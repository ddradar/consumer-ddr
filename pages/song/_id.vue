<template>
  <section v-if="info" class="section">
    <h1 class="title">{{ info.name }} / {{ info.artist }}</h1>
    <h2 class="subtitle">BPM {{ info.bpm }}</h2>
    <div class="buttons">
      <OButton
        v-for="id in info.seriesList"
        :key="id"
        tag="NuxtLink"
        size="small"
        :variant="getClass(id)"
        :to="`/series/${id}/`"
      >
        {{ id }}
      </OButton>
    </div>
    <OTable :data="info.charts" striped>
      <OTableColumn v-slot="props" field="playStyle" label="PlayStyle">
        {{ props.row.playStyle }}
      </OTableColumn>
      <OTableColumn v-slot="props" field="difficulty" label="Difficulty">
        {{ props.row.difficulty }}
      </OTableColumn>
      <OTableColumn v-slot="props" field="level" label="Level">
        {{ Object.values(props.row.levels).join(', ') }}
      </OTableColumn>
    </OTable>
  </section>
</template>

<script lang="ts">
import { Context } from '@nuxt/types'
import { Component, Vue } from 'nuxt-property-decorator'
import { MetaInfo } from 'vue-meta'

import { Chart, Song } from '~/types/song'

type ComplexChart = Omit<Chart, 'level'> & {
  levels: { [key in string]: number | '10+' | '?' }
}

type SongInfo = Omit<Song, 'series' | 'charts'> & {
  charts: ComplexChart[]
  seriesList: string[]
}

@Component
export default class SongDetailPage extends Vue {
  info: SongInfo | null = null

  async asyncData({ params, $content }: Pick<Context, 'params' | '$content'>) {
    const songs = await $content({ deep: true })
      .where({ slug: params.id })
      .fetch<Song>()
    const info = [songs].flat().reduce((prev, current) => {
      if (!prev) {
        return {
          name: current.name,
          artist: current.artist,
          bpm: current.bpm,
          charts: current.charts.map<ComplexChart>((c) => {
            const levels: { [key in string]: number | '10+' | '?' } = {}
            levels[current.series] = c.level
            return {
              playStyle: c.playStyle,
              difficulty: c.difficulty,
              levels
            }
          }),
          seriesList: []
        } as SongInfo
      }
      prev.seriesList.push(current.series)
      for (const chart of current.charts) {
        const existsChart = prev.charts.find(
          (c) =>
            c.playStyle === chart.playStyle && c.difficulty === chart.difficulty
        )
        if (existsChart) {
          existsChart.levels[current.series] = chart.level
        } else {
          const levels: { [key in string]: number | '10+' | '?' } = {}
          levels[current.series] = chart.level
          prev.charts.push({
            playStyle: chart.playStyle,
            difficulty: chart.difficulty,
            levels
          })
        }
      }
      return prev
    }, undefined as SongInfo | undefined)
    return { info }
  }

  head(): MetaInfo {
    return {
      title: this.info?.name
    }
  }

  getClass(id: string) {
    const index = this.info!.seriesList.findIndex((s) => s === id)
    const classList = ['info', 'success', 'danger', 'warning', 'dark']
    return classList[index % classList.length]
  }
}
</script>
