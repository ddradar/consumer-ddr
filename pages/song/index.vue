<template>
  <section class="section">
    <h1 class="title">Song List</h1>

    <OTable :data="songList" per-page="20" striped narrowed paginated>
      <OTableColumn v-slot="props" field="name" label="Name" searchable>
        <NuxtLink class="is-size-6-mobile" :to="`/song/${props.row.slug}/`">
          {{ props.row.name }}
        </NuxtLink>
      </OTableColumn>
      <OTableColumn v-slot="props" field="artist" label="Artist" searchable>
        <span class="is-size-6-mobile">{{ props.row.artist }}</span>
      </OTableColumn>
      <OTableColumn v-slot="props" field="bpm" label="BPM">
        {{ props.row.bpm }}
      </OTableColumn>
      <OTableColumn v-slot="props" label="Series">
        <div class="buttons">
          <OButton
            v-for="id in props.row.seriesList"
            :key="id"
            tag="NuxtLink"
            size="small"
            :variant="getClass(id)"
            :to="`/series/${id}/`"
          >
            {{ id }}
          </OButton>
        </div>
      </OTableColumn>

      <template #empty>
        <section class="section">
          <div class="content has-text-grey has-text-centered">
            <p>Nothing here.</p>
          </div>
        </section>
      </template>
    </OTable>
  </section>
</template>

<script lang="ts">
import { Context } from '@nuxt/types'
import { Component, Vue } from 'nuxt-property-decorator'
import { MetaInfo } from 'vue-meta'

import { Song } from '~/types/song'

type SongListData = Omit<Song, 'series' | 'charts'> & {
  seriesList: string[]
}

@Component
export default class SeriesDetailPage extends Vue {
  songList: SongListData[] = []
  seriesList: string[] = []

  readonly head: MetaInfo = {
    title: 'Song List'
  }

  async asyncData({ $content }: Pick<Context, '$content'>) {
    const songs = await $content({ deep: true })
      .where({ extension: { $eq: '.json' } })
      .without('charts')
      .fetch<Omit<Song, 'charts'>>()
    const songList = [songs].flat().reduce((prev, current) => {
      if (!prev) {
        return [
          {
            slug: current.slug,
            name: current.name,
            artist: current.artist,
            bpm: current.bpm,
            seriesList: [current.series]
          }
        ]
      }
      if (!prev.some((s) => s.slug === current.slug)) {
        prev.push({
          slug: current.slug,
          name: current.name,
          artist: current.artist,
          bpm: current.bpm,
          seriesList: [current.series]
        })
      }
      return prev
    }, undefined as SongListData[] | undefined)

    const seriesList = (await $content({ deep: true })
      .where({ extension: { $eq: '.md' } })
      .sortBy('launched')
      .only('slug')
      .fetch<string>()) as string[]

    return { songList, seriesList }
  }

  getClass(id: string) {
    const index = this.seriesList!.findIndex((s) => s === id)
    const classList = ['info', 'success', 'danger', 'warning', 'dark']
    return classList[index % classList.length]
  }
}
</script>
