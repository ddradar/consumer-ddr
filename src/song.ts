const basicNames = ['BASIC', 'NORMAL', 'LIGHT', '楽']
const difficultNames = ['ANOTHER', 'DIFFICULT', 'TRICK', '踊', 'STANDARD']
const expertNames = ['MANIAC', 'SSR', 'EXPERT', '激', 'HEAVY']
const beginnerNames = ['BEGINNER']

export function normalizeDifficulty(difficultyName: string) {
  const name = difficultyName.toUpperCase()
  if (basicNames.includes(name)) return 'BASIC'
  if (difficultNames.includes(name)) return 'DIFFICULT'
  if (expertNames.includes(name)) return 'EXPERT'
  if (beginnerNames.includes(name)) return 'BEGINNER'
  return 'CHALLENGE'
}
