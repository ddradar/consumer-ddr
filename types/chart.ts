type PlayStyle = 'SINGLE' | 'DOUBLE' | 'COUPLE' | '6-PANELS' | '3-PANELS'

export type Chart = {
  songId: string
  playStyle: PlayStyle
  difficulty: number
  level: number | '10+'
}
