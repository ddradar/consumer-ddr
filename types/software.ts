type Region = 'JP' | 'US' | 'EU'

type Platform = 'Play Station' | 'Play Station 2' | 'GAME BOY' | 'Wii'

export type Software = {
  id: string
  name: string
  region: Region
  platform: Platform
  launched: Date | null
}
