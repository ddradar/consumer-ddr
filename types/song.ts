export type Song = {
  slug?: string
  series: string
  name: string
  artist: string
  bpm: number | string
  charts: Chart[]
}

export type Chart = {
  playStyle: PlayStyle
  difficulty: number
  level: number | '10+' | '?'
}

export type PlayStyle =
  | 'SINGLE'
  | 'DOUBLE'
  | 'COUPLE'
  | 'UNISON'
  | '6-PANELS'
  | '3-PANELS'
  | 'STEP BATTLE'

export function normalizeDifficulty(difficultyName: string) {
  const name = difficultyName.toUpperCase()
  if (['BASIC', 'NORMAL', 'LIGHT'].includes(name)) return 'BASIC'
  if (['ANOTHER', 'DIFFICULT', 'TRICK'].includes(name)) return 'DIFFICULT'
  if (['MANIAC', 'SSR', 'EXPERT'].includes(name)) return 'EXPERT'
  if (['BEGINNER'].includes(name)) return 'BEGINNER'
  return 'CHALLENGE'
}
