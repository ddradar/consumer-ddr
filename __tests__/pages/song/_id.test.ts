import { mount, RouterLinkStub, shallowMount, Wrapper } from '@vue/test-utils'

import { createVue } from '~/__tests__/utils'
import SongDetail from '~/pages/song/_id.vue'
import { Chart, Song } from '~/types/song'

type ComplexChart = Omit<Chart, 'level'> & {
  levels: { [key in string]: number | '10+' | '?' }
}

type SongInfo = Omit<Song, 'series' | 'charts'> & {
  charts: ComplexChart[]
  seriesList: string[]
}

const localVue = createVue()

describe('pages/song/_id.vue', () => {
  let wrapper: Wrapper<SongDetail>
  const info: SongInfo = {
    slug: 'boys-smiledk',
    name: 'BOYS',
    artist: 'smile. dk',
    bpm: 138,
    seriesList: ['1st-jp', '2nd-remix-jp'],
    charts: [
      {
        playStyle: 'SINGLE',
        difficulty: 1,
        levels: { '1st-jp': 3, '2nd-remix-jp': 3 }
      },
      {
        playStyle: 'SINGLE',
        difficulty: 2,
        levels: { '1st-jp': 4, '2nd-remix-jp': 4 }
      },
      {
        playStyle: 'SINGLE',
        difficulty: 3,
        levels: { '1st-jp': 7, '2nd-remix-jp': 7 }
      },
      {
        playStyle: 'SINGLE',
        difficulty: 1,
        levels: { '1st-jp': 4, '2nd-remix-jp': 4 }
      },
      {
        playStyle: 'SINGLE',
        difficulty: 2,
        levels: { '1st-jp': 5, '2nd-remix-jp': 5 }
      }
    ]
  }
  const stubs = { NuxtLink: RouterLinkStub }
  const data = () => ({ info })

  beforeEach(() => {
    wrapper = shallowMount(SongDetail, { localVue, data, stubs })
  })

  test('renders correctly', async () => {
    const wrapper = mount(SongDetail, { localVue, data, stubs })
    await wrapper.vm.$nextTick()
    expect(wrapper.element).toMatchSnapshot()
  })
  test('renders blank', () => {
    const wrapper = shallowMount(SongDetail, { localVue })
    expect(wrapper.html()).toBe('')
  })

  describe('head()', () => {
    test('returns info.name', () => {
      // Arrange
      const head = wrapper.vm.$options.head as any

      // Act & Assert
      expect(head.call(wrapper.vm)).toStrictEqual({ title: info.name })
    })
  })
})
