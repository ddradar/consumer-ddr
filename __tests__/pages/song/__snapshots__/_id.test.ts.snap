// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/song/_id.vue renders correctly 1`] = `
<section
  class="section"
>
  <h1
    class="title"
  >
    BOYS / smile. dk
  </h1>
   
  <h2
    class="subtitle"
  >
    BPM 138
  </h2>
   
  <div
    class="buttons"
  >
    <a
      class="button is-small is-info"
    >
      <span
        class="button-wrapper"
      >
        <!---->
        <span
          class=""
        >
          
      1st-jp
    
        </span>
        <!---->
      </span>
    </a>
    <a
      class="button is-small is-success"
    >
      <span
        class="button-wrapper"
      >
        <!---->
        <span
          class=""
        >
          
      2nd-remix-jp
    
        </span>
        <!---->
      </span>
    </a>
  </div>
   
  <div
    class="b-table"
  >
    <div
      style="display: none;"
    >
      <span>
        PlayStyle
      </span>
       
      <span>
        Difficulty
      </span>
       
      <span>
        Level
      </span>
    </div>
    <!---->
    <!---->
    <div
      class="table-wrapper"
    >
      <table
        class="table is-striped"
      >
        <!---->
        <thead>
          <tr>
            <!---->
            <!---->
            <th
              class=""
              draggable="false"
            >
              <span>
                 PlayStyle 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                 Difficulty 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                 Level 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <!---->
          </tr>
          <!---->
          <!---->
        </thead>
        <tbody>
          <tr
            class=""
            draggable="false"
          >
            <!---->
            <!---->
            <td
              class=""
              data-label="PlayStyle"
            >
              
      SINGLE
    
            </td>
            <td
              class=""
              data-label="Difficulty"
            >
              
      1
    
            </td>
            <td
              class=""
              data-label="Level"
            >
              
      3, 3
    
            </td>
            <!---->
          </tr>
          <transition-stub
            name=""
          >
            <!---->
          </transition-stub>
          <!---->
          <tr
            class=""
            draggable="false"
          >
            <!---->
            <!---->
            <td
              class=""
              data-label="PlayStyle"
            >
              
      SINGLE
    
            </td>
            <td
              class=""
              data-label="Difficulty"
            >
              
      2
    
            </td>
            <td
              class=""
              data-label="Level"
            >
              
      4, 4
    
            </td>
            <!---->
          </tr>
          <transition-stub
            name=""
          >
            <!---->
          </transition-stub>
          <!---->
          <tr
            class=""
            draggable="false"
          >
            <!---->
            <!---->
            <td
              class=""
              data-label="PlayStyle"
            >
              
      SINGLE
    
            </td>
            <td
              class=""
              data-label="Difficulty"
            >
              
      3
    
            </td>
            <td
              class=""
              data-label="Level"
            >
              
      7, 7
    
            </td>
            <!---->
          </tr>
          <transition-stub
            name=""
          >
            <!---->
          </transition-stub>
          <!---->
          <tr
            class=""
            draggable="false"
          >
            <!---->
            <!---->
            <td
              class=""
              data-label="PlayStyle"
            >
              
      SINGLE
    
            </td>
            <td
              class=""
              data-label="Difficulty"
            >
              
      1
    
            </td>
            <td
              class=""
              data-label="Level"
            >
              
      4, 4
    
            </td>
            <!---->
          </tr>
          <transition-stub
            name=""
          >
            <!---->
          </transition-stub>
          <!---->
          <tr
            class=""
            draggable="false"
          >
            <!---->
            <!---->
            <td
              class=""
              data-label="PlayStyle"
            >
              
      SINGLE
    
            </td>
            <td
              class=""
              data-label="Difficulty"
            >
              
      2
    
            </td>
            <td
              class=""
              data-label="Level"
            >
              
      5, 5
    
            </td>
            <!---->
          </tr>
          <transition-stub
            name=""
          >
            <!---->
          </transition-stub>
          <!---->
          <!---->
        </tbody>
        <!---->
      </table>
      <transition-stub
        name="fade"
      >
        <!---->
      </transition-stub>
    </div>
    <!---->
  </div>
</section>
`;
